tower-http/src/auth/add_authorization.rs

```diff
@@ -0,0 +1,213 @@
+//! Add authorization to requests using the [`Authorization`] header.
+//!
+//! [`Authorization`]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization
+//!
+//! # Example
+//!
+//! ```
+//! use tower_http::auth::AddAuthorizationLayer;
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::AUTHORIZATION};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//! # async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//! #     Ok(Response::new(Body::empty()))
+//! # }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let service_that_requires_auth = tower_http::auth::RequireAuthorization::basic(
+//! #     tower::service_fn(handle),
+//! #     "username",
+//! #     "password",
+//! # );
+//! let mut client = ServiceBuilder::new()
+//!     // Use basic auth with the given username and password
+//!     .layer(AddAuthorizationLayer::basic("username", "password"))
+//!     .service(service_that_requires_auth);
+//!
+//! // Make a request, we don't have to add the `Authorization` header manually
+//! let response = client
+//!     .ready()
+//!     .await?
+//!     .call(Request::new(Body::empty()))
+//!     .await?;
+//!
+//! assert_eq!(StatusCode::OK, response.status());
+//! # Ok(())
+//! # }
+//! ```
+
+use http::{HeaderValue, Request};
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`AddAuthorization`] which adds authorization to all requests using the
+/// [`Authorization`] header.
+///
+/// See the [module docs](crate::auth::add_authorization) for an example.
+///
+/// You can also use [`SetRequestHeader`] if you have a use case that isn't supported by this
+/// middleware.
+///
+/// [`Authorization`]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization
+/// [`SetRequestHeader`]: crate::set_header::SetRequestHeader
+#[derive(Debug, Clone)]
+pub struct AddAuthorizationLayer {
+    value: HeaderValue,
+}
+
+impl AddAuthorizationLayer {
+    /// Authorize requests using a username and password pair.
+    ///
+    /// The `Authorization` header will be set to `Basic {credentials}` where `credentials` is
+    /// `base64_encode("{username}:{password}")`.
+    ///
+    /// Since the username and password is sent in clear text it is recommended to use HTTPS/TLS
+    /// with this method. However use of HTTPS/TLS is not enforced by this middleware.
+    pub fn basic(username: &str, password: &str) -> Self {
+        let encoded = base64::encode(format!("{}:{}", username, password));
+        let value = HeaderValue::from_str(&format!("Basic {}", encoded)).unwrap();
```

Comment:
```suggestion
        let mut value = HeaderValue::from_str(&format!("Basic {}", encoded)).unwrap();
        value.set_sensitive(true);
```

Reply:
I think adding a doc comment about the value not being marked as sensitive and linking to https://docs.rs/tower-http/0.1.0/tower_http/sensitive_headers/struct.SetSensitiveRequestHeaders.html is a bit nicer.

Reply:
Yeah, I thought that might be the reason. Is there a reason not to set sensitive?

In `kube`, we provide a default stack based on config, and it has an optional authorization layer (`Option<Either<AddAuthorizationLayer, SomeOtherAuthLayer>>`) applied.

```rust
let stack = ServiceBuilder::new()
    .option_layer(maybe_auth) // depends on the config
    // ... layers
    .layer(DecompressionLayer::new())
    .layer(TraceLayer::new_for_http()) // BTW, thanks for these :)
    .build(hyper_client);
```
I can add `.layer(SetSensitiveHeadersLayer::new(once(AUTHORIZATION)))`, but feels awkward since the authorization header is optional. It's not a huge deal, so I'll switch to whatever `tower-http` provides, but it was a bit unexpected.

Reply:
Whoa so cool that you're using tower-http in kube ❤️

Having thought a bit more about this I think a builder style config method would probably nice. So something like:

```rust
.layer(AddAuthorizationLayer::basic("username", "password").as_sensitive(true))
```

Not sure that `as_sensitive` is the best name so suggestions welcome.

Reply:
Yeah, that would be nice :+1: 
`as_sensitive` looks fine to me. But just `sensitive()` (no bool) works too.

Reply:
I think the method taking a bool is nice because you can more easily set it according to some env var or something. Like `.as_sensitive(env::var("LOCAL").is_ok())`.

I've added it in 94a7cb012d0be.